src/logging.cpp

```diff
@@ -29,45 +36,144 @@ bl::sources::severity_logger<int> warning(3);  // Strange events
-    << "        -p | Enable/Disable UPnP"sv << std::endl
-    << std::endl;
-}
+BOOST_LOG_ATTRIBUTE_KEYWORD(severity, "Severity", int)
+
+namespace logging {
+  /**
+   * @brief Initialize the logging system.
+   *
+   * EXAMPLES:
+   * ```cpp
+   * log_init();
+   * ```
+   */
+  void
+  init() {
+    if (config::sunshine.min_log_level >= 1) {
+      av_log_set_level(AV_LOG_QUIET);
+    }
+    else {
+      av_log_set_level(AV_LOG_DEBUG);
+    }
+    av_log_set_callback([](void *ptr, int level, const char *fmt, va_list vl) {
+      static int print_prefix = 1;
+      char buffer[1024];
+
+      av_log_format_line(ptr, level, fmt, vl, buffer, sizeof(buffer), &print_prefix);
+      if (level <= AV_LOG_ERROR) {
+        // We print AV_LOG_FATAL at the error level. FFmpeg prints things as fatal that
+        // are expected in some cases, such as lack of codec support or similar things.
+        BOOST_LOG(error) << buffer;
+      }
+      else if (level <= AV_LOG_WARNING) {
+        BOOST_LOG(warning) << buffer;
+      }
+      else if (level <= AV_LOG_INFO) {
+        BOOST_LOG(info) << buffer;
+      }
+      else if (level <= AV_LOG_VERBOSE) {
+        // AV_LOG_VERBOSE is less verbose than AV_LOG_DEBUG
+        BOOST_LOG(debug) << buffer;
+      }
+      else {
+        BOOST_LOG(verbose) << buffer;
+      }
+    });
+
+    sink = boost::make_shared<text_sink>();
+
+    boost::shared_ptr<std::ostream> stream { &std::cout, NoDelete {} };
+    sink->locked_backend()->add_stream(stream);
+    sink->locked_backend()->add_stream(boost::make_shared<std::ofstream>(config::sunshine.log_file));
+    sink->set_filter(severity >= config::sunshine.min_log_level);
+
+    sink->set_formatter([message = "Message"s, severity = "Severity"s](const bl::record_view &view, bl::formatting_ostream &os) {
+      constexpr int DATE_BUFFER_SIZE = 21 + 2 + 1;  // Full string plus ": \0"
+
+      auto log_level = view.attribute_values()[severity].extract<int>().get();
+
+      std::string_view log_type;
+      switch (log_level) {
+        case 0:
+          log_type = "Verbose: "sv;
+          break;
+        case 1:
+          log_type = "Debug: "sv;
+          break;
+        case 2:
+          log_type = "Info: "sv;
+          break;
+        case 3:
+          log_type = "Warning: "sv;
+          break;
+        case 4:
+          log_type = "Error: "sv;
+          break;
+        case 5:
+          log_type = "Fatal: "sv;
+          break;
+      };
+
+      char _date[DATE_BUFFER_SIZE];
+      std::time_t t = std::time(nullptr);
+      strftime(_date, DATE_BUFFER_SIZE, "[%Y:%m:%d:%H:%M:%S]: ", std::localtime(&t));
+
+      os << _date << log_type << view.attribute_values()[message].extract<std::string>();
+    });
+
+    // Flush after each log record to ensure log file contents on disk isn't stale.
+    // This is particularly important when running from a Windows service.
+    sink->locked_backend()->auto_flush(true);
```

Comment:
If this is enabled, why are we even flushing manually?

Reply:
I'm not sure, perhaps the manual flush was just left over from before auto_flush was added?Maybe @cgutman has some idea?